### PR TITLE
kubeadm: graduate control plane prepare phase

### PR DESCRIPTION
**Pull request by ereslibre**
**What this PR does / why we need it**:
Graduate control plane prepare phase

**Which issue(s) this PR fixes**:
Refs https://github.com/kubernetes/kubeadm/issues/1204

**Does this PR introduce a user-facing change?**:
```release-note
NONE
```

/kind feature

### CODE REVIEW

**Comment by ereslibre** [comment]
Submitting for early feedback. cc/ @fabriziopandini @neolit123

**Review by yagonobre — COMMENTED** [review]
Thanks @ereslibre 
/ok-to-test
/kind cleanup
/priority important-soon

File: cmd/kubeadm/app/cmd/phases/controlplanepreparejoin.go
Comment (yagonobre) [inline cmd/kubeadm/app/cmd/phases/controlplanepreparejoin.go]: I guess that we just need to declare the functions instead of returning this.

File: cmd/kubeadm/app/cmd/phases/controlplanepreparejoin.go
Comment (yagonobre) [inline cmd/kubeadm/app/cmd/phases/controlplanepreparejoin.go]: We can call `data.Cfg()` direct here.

File: cmd/kubeadm/app/cmd/phases/controlplanepreparejoin.go
Comment (yagonobre) [inline cmd/kubeadm/app/cmd/phases/controlplanepreparejoin.go]: We can use `data.Cfg()` here as well.

File: cmd/kubeadm/app/cmd/phases/controlplanepreparejoin.go
Comment (yagonobre) [inline cmd/kubeadm/app/cmd/phases/controlplanepreparejoin.go]: Can we wrap the error or just use `return certsphase.CreatePKIAssets(cfg)`?

File: cmd/kubeadm/app/cmd/phases/controlplanepreparejoin.go
Comment (yagonobre) [inline cmd/kubeadm/app/cmd/phases/controlplanepreparejoin.go]: Maybe we can use `return etcdphase.CheckLocalEtcdClusterStatus(client, cfg)`

**Review by yagonobre — COMMENTED** [review]
@ereslibre SGTM, I have added few comments.

File: cmd/kubeadm/app/cmd/phases/controlplanepreparejoin.go
Comment (MalloZup) [inline cmd/kubeadm/app/cmd/phases/controlplanepreparejoin.go]: we can return it withou this var	`return workflow.Phase{`

File: cmd/kubeadm/app/cmd/phases/controlplanepreparejoin.go
Comment (MalloZup) [inline cmd/kubeadm/app/cmd/phases/controlplanepreparejoin.go]: idem as previous comment

File: cmd/kubeadm/app/cmd/phases/controlplanepreparejoin.go
Comment (MalloZup) [inline cmd/kubeadm/app/cmd/phases/controlplanepreparejoin.go]: idem :champagne:

File: cmd/kubeadm/app/cmd/phases/controlplanepreparejoin.go
Comment (MalloZup) [inline cmd/kubeadm/app/cmd/phases/controlplanepreparejoin.go]: idem :champagne:

File: cmd/kubeadm/app/cmd/phases/controlplanepreparejoin.go
Comment (MalloZup) [inline cmd/kubeadm/app/cmd/phases/controlplanepreparejoin.go]: afaik here without `(). so `Run:   runKubeconfigPhaseLocal`,

**Review by MalloZup — COMMENTED** [review]
:unicorn:

File: cmd/kubeadm/app/cmd/phases/controlplanepreparejoin.go
Comment (ereslibre) [reply cmd/kubeadm/app/cmd/phases/controlplanepreparejoin.go]: Ugh, yes.

File: cmd/kubeadm/app/cmd/phases/controlplanepreparejoin.go
Comment (ereslibre) [reply cmd/kubeadm/app/cmd/phases/controlplanepreparejoin.go]: I would prefer not to do this here, and only return in case of error in this case, just to keep the flow in case this function gets extended after this block.

**Review by yagonobre — APPROVED** [review]
@ereslibre thanks for the fast interaction.
/lgtm

File: cmd/kubeadm/app/cmd/phases/controlplanepreparejoin.go
Comment (MalloZup) [inline cmd/kubeadm/app/cmd/phases/controlplanepreparejoin.go]: we might add `example` section

File: cmd/kubeadm/app/cmd/phases/controlplanepreparejoin.go
Comment (MalloZup) [inline cmd/kubeadm/app/cmd/phases/controlplanepreparejoin.go]: imho for each phase, seen as `atomic` User operation, we should add an `example` section. where usere can have example to call the phase.

File: cmd/kubeadm/app/cmd/phases/controlplanepreparejoin.go
Comment (MalloZup) [inline cmd/kubeadm/app/cmd/phases/controlplanepreparejoin.go]: `nit` we could name it `joinData`, might have more context. (but i'm ok for both)

File: cmd/kubeadm/app/cmd/phases/controlplanepreparejoin.go
Comment (MalloZup) [inline cmd/kubeadm/app/cmd/phases/controlplanepreparejoin.go]: we might have coherence/guideline about errors. `should we wrap all errors`  or return them?

Comment (MalloZup) [reply cmd/kubeadm/app/cmd/phases/controlplanepreparejoin.go]: e.g should we add `[join/phaseName]` in the error string for being more precise?

**Review by MalloZup — COMMENTED** [review]
added some comments, which some of them are also open question . thx @ereslibre :robot:

File: cmd/kubeadm/app/cmd/phases/controlplanepreparejoin.go
Comment (MalloZup) [reply cmd/kubeadm/app/cmd/phases/controlplanepreparejoin.go]: https://github.com/kubernetes/kubernetes/blob/master/cmd/kubeadm/app/cmd/phases/preflight.go#L54

**Comment by MalloZup** [comment]
@ereslibre @yagonobre afaik 

`Does this PR introduce a user-facing change?:`

Imho we are adding a UX change. Namely we are introducing per command line the possibility to have phases like

`kubeadm join phase ...`  which was not there before afaik

File: cmd/kubeadm/app/cmd/phases/controlplanepreparejoin.go
Comment (yagonobre) [reply cmd/kubeadm/app/cmd/phases/controlplanepreparejoin.go]: I guess that `[phaseName]` is a must have

**Comment by yagonobre** [comment]
@MalloZup we do the changes in a couples of pr, so we can left the release note as none and add a single note on the release time.

File: cmd/kubeadm/app/cmd/phases/controlplanepreparejoin.go
Comment (ereslibre) [reply cmd/kubeadm/app/cmd/phases/controlplanepreparejoin.go]: Do we really need it for the subphases?

File: cmd/kubeadm/app/cmd/phases/controlplanepreparejoin.go
Comment (ereslibre) [reply cmd/kubeadm/app/cmd/phases/controlplanepreparejoin.go]: I'm not sure if they are not having explicit options TBH.

File: cmd/kubeadm/app/cmd/phases/controlplanepreparejoin.go
Comment (ereslibre) [reply cmd/kubeadm/app/cmd/phases/controlplanepreparejoin.go]: I am keeping `data` as we have been using for the rest of the init subphases.

File: cmd/kubeadm/app/cmd/phases/controlplanepreparejoin.go
Comment (MalloZup) [reply cmd/kubeadm/app/cmd/phases/controlplanepreparejoin.go]: afaik `subphases` can be called in a `public` manner. If they are like an `internal` function i agree with you.

**Review by MalloZup — APPROVED** [review]
a part of the small comments looks good

**Comment by ereslibre** [comment]
/hold

Need to revisit the PR.

**Comment by ereslibre** [comment]
/hold

I want to improve this PR to treat every component as we do with `init` in the `manifests` subphase. Will ask for review when I'm finished.

**Comment by ereslibre** [comment]
/hold cancel

I added the manifests as separate subphases so the UX is similar to `kubeadm init`.

Also, the panic and the type assertion will be fixed on https://github.com/kubernetes/kubernetes/pull/73348

**Comment by ereslibre** [comment]
Rebased on top of https://github.com/kubernetes/kubernetes/pull/73348

File: cmd/kubeadm/app/cmd/phases/controlplanepreparejoin.go
Comment (neolit123) [reply cmd/kubeadm/app/cmd/phases/controlplanepreparejoin.go]: for some phases we don't have examples, so we are not consistent.
for instance the `init phase certs` phase doesn't have them
i'm OK with not having examples for this phase.

possibly we can update them all in a single PR later.
but my vote would be to not have examples because our commands are not very complicated.

File: cmd/kubeadm/app/cmd/phases/controlplanepreparejoin.go
Comment (neolit123) [inline cmd/kubeadm/app/cmd/phases/controlplanepreparejoin.go]: this should have been part of @MalloZup 's phase (preflight)
looks like we've missed it.

https://github.com/kubernetes/kubeadm/issues/1204
> etcd check (if initConfiguration.Etcd.External == nil from PrepareForHostingControlPlane)

@ereslibre could you please move it as part of this PR to preflight.go?

File: cmd/kubeadm/app/cmd/phases/controlplanepreparejoin.go
Comment (neolit123) [inline cmd/kubeadm/app/cmd/phases/controlplanepreparejoin.go]: we need to use `ControlPlanePrepare` in the names for these functions.

e,g. for this one we have a conflict with: https://github.com/kubernetes/kubernetes/blob/master/cmd/kubeadm/app/cmd/phases/certs.go#L86

`newControlPlanePrepareCertsSubphase`
`newControlPlanePrepareKubeconfigSubphase`
...
same for the run method names.

this isn't great looking but at least it makes it clear what this function is doing given we ended up in the same namespaces.

we should have used a separate namespace anyhow.

File: cmd/kubeadm/app/cmd/phases/controlplanepreparejoin.go
Comment (neolit123) [inline cmd/kubeadm/app/cmd/phases/controlplanepreparejoin.go]: our pattern in other files is to use the name of the parent phase in these messages - so `control-plane-prepare`.
e.g. https://github.com/kubernetes/kubernetes/blob/master/cmd/kubeadm/app/cmd/phases/certs.go#L86
[1]

File: cmd/kubeadm/app/cmd/phases/controlplanepreparejoin.go
Comment (neolit123) [inline cmd/kubeadm/app/cmd/phases/controlplanepreparejoin.go]: [1]

File: cmd/kubeadm/app/cmd/phases/controlplanepreparejoin.go
Comment (neolit123) [inline cmd/kubeadm/app/cmd/phases/controlplanepreparejoin.go]: as mentioned we need this in preflight instead.

File: cmd/kubeadm/app/cmd/phases/controlplanepreparejoin.go
Comment (neolit123) [inline cmd/kubeadm/app/cmd/phases/controlplanepreparejoin.go]: let's add something in these lines:
`fmt.Println("[control-plane-prepare] Generating kubeconfig files")`

File: cmd/kubeadm/app/cmd/phases/controlplanepreparejoin.go
Comment (neolit123) [inline cmd/kubeadm/app/cmd/phases/controlplanepreparejoin.go]: still print a message that we are performing an action (preflight)?
`fmt.Println("[preflight] Checking if etcd cluster is healthy")`

File: cmd/kubeadm/app/cmd/phases/util.go
Comment (neolit123) [inline cmd/kubeadm/app/cmd/phases/util.go]: i'm pretty sure this will cause a build problem given `getPhaseDescription` is already declared in the same package?
https://github.com/kubernetes/kubernetes/blob/master/cmd/kubeadm/app/cmd/phases/controlplane.go

we should remove it from the above file if it is now in utils.go

**Review by neolit123 — COMMENTED** [review]
@ereslibre thanks for the PR.
added some comments.

probably the biggest topic is to move the etcd check in preflight.
we should have done that in @MalloZup 's PR.

sorry for not catching that.

File: cmd/kubeadm/app/cmd/phases/util.go
Comment (ereslibre) [reply cmd/kubeadm/app/cmd/phases/util.go]: In this PR I'm removing it from `controlplane.go` (renamed to `controlplaneinit.go`) and moved it to `util.go`, so there's no problem with that.

That being said, maybe `util.go` is too generic and we might want a `controlplaneutil.go` even in the same package just for better organization?

File: cmd/kubeadm/app/cmd/phases/controlplanepreparejoin.go
Comment (ereslibre) [reply cmd/kubeadm/app/cmd/phases/controlplanepreparejoin.go]: I will adapt, thanks!

File: cmd/kubeadm/app/cmd/phases/controlplanepreparejoin.go
Comment (ereslibre) [reply cmd/kubeadm/app/cmd/phases/controlplanepreparejoin.go]: Every component will write `[kubeconfig] Writing %q kubeconfig file`, should we still write a generic message before those?

File: cmd/kubeadm/app/cmd/phases/util.go
Comment (neolit123) [reply cmd/kubeadm/app/cmd/phases/util.go]: > In this PR I'm removing it from controlplane.go
i guess i missed that.

`util.go` seems OK for now. we can do some of these refactors later.

File: cmd/kubeadm/app/cmd/phases/controlplanepreparejoin.go
Comment (neolit123) [reply cmd/kubeadm/app/cmd/phases/controlplanepreparejoin.go]: might be a problem, yes...
please check, but if that is the case we should leave it here for now.

File: cmd/kubeadm/app/cmd/phases/controlplanepreparejoin.go
Comment (fabriziopandini) [inline cmd/kubeadm/app/cmd/phases/controlplanepreparejoin.go]: I don't think that discovery flags are relevant for this phase. Could you kindly check and remove what is not relevant for this phase

File: cmd/kubeadm/app/cmd/phases/controlplanepreparejoin.go
Comment (fabriziopandini) [reply cmd/kubeadm/app/cmd/phases/controlplanepreparejoin.go]: 🤔 
I just checked this and this is applied consistently to `fmt.Printf`, but not to errors.
Thinking about this, errors should be clean because the error could be concatenated other errors

File: cmd/kubeadm/app/cmd/phases/controlplanepreparejoin.go
Comment (fabriziopandini) [inline cmd/kubeadm/app/cmd/phases/controlplanepreparejoin.go]: IMO all the paths should be part of the joinData (see init as a reference)
Opinions?

File: cmd/kubeadm/app/cmd/phases/controlplanepreparejoin.go
Comment (fabriziopandini) [inline cmd/kubeadm/app/cmd/phases/controlplanepreparejoin.go]: Client should be part of JoinData and be implemented as a singleton

File: cmd/kubeadm/app/cmd/phases/controlplanepreparejoin.go
Comment (fabriziopandini) [reply cmd/kubeadm/app/cmd/phases/controlplanepreparejoin.go]: I'm not sure this can easily moved on preflight due to the dependency defined above, let's chat about this

**Review by fabriziopandini — CHANGES_REQUESTED** [review]
@ereslibre Thanks for this PR!
This is already a good starting point, even if an iteration is required to address some points:
- file organization (i see a different approach here and in preflight)
- the etcd check
What about trying to organize a short meeting to discuss things F2F?

File: cmd/kubeadm/app/cmd/phases/controlplanepreparejoin.go
Comment (ereslibre) [reply cmd/kubeadm/app/cmd/phases/controlplanepreparejoin.go]: `TokenDiscovery` needs to be passed for the validation to succeed if you don't use a config file andn point directly to the master as an unnamed arg. I didn't want to mess too much with the validation logic in this step, so I went with the straightforward approach: if you call to the subphase directly you are able to pass the server as an unnamed argument, and the flags required for the validation to succeed.

File: cmd/kubeadm/app/cmd/phases/controlplanepreparejoin.go
Comment (ereslibre) [reply cmd/kubeadm/app/cmd/phases/controlplanepreparejoin.go]: This was like this in the join logic before and I felt like a refactor of this specifically in this PR was not going to be a good idea, and I left it as it was. Same for certificate location, on join it's not possible to configure it AFAIK.

File: cmd/kubeadm/app/cmd/phases/controlplanepreparejoin.go
Comment (ereslibre) [reply cmd/kubeadm/app/cmd/phases/controlplanepreparejoin.go]: I will refactor this.

Comment (ereslibre) [reply cmd/kubeadm/app/cmd/phases/controlplanepreparejoin.go]: On a second pass we are creating three kind of Clientsets: admin, bootstrap kubelet and kubelet.

The operation of creating the clientset should be cheap (i.e. no remote actions), I wonder if having this as a singleton backed up by a map is making us gain something?

Comment (ereslibre) [reply cmd/kubeadm/app/cmd/phases/controlplanepreparejoin.go]: Added a second commit (to be squashed when we are done with review). Is this what you mean?

**Comment by ereslibre** [comment]
@fabriziopandini @neolit123 I just reorganized the code to have everything in `controlplane.go` as with preflight. I would prefer a slightly different approach splitting in different files and potentially packages but for a first iteration I added a third commit (will be squashed if we agree) that reorganizes the code. Thanks!

File: cmd/kubeadm/app/cmd/phases/controlplanepreparejoin.go
Comment (ereslibre) [reply cmd/kubeadm/app/cmd/phases/controlplanepreparejoin.go]: So, for checking the etcd health we are using the etcd healthcheck certificate and key that are not copied among masters, but generated automatically on every master. So I'd say we should keep this subphase after the PKI generation in this phase.

File: cmd/kubeadm/app/cmd/phases/controlplanepreparejoin.go
Comment (ereslibre) [reply cmd/kubeadm/app/cmd/phases/controlplanepreparejoin.go]: Updated, I think we should probably refactor after all this work to use different packages.

File: cmd/kubeadm/app/cmd/phases/controlplanepreparejoin.go
Comment (ereslibre) [reply cmd/kubeadm/app/cmd/phases/controlplanepreparejoin.go]: I'll resolve this one in favour of https://github.com/kubernetes/kubernetes/pull/73145#discussion_r252666255 discussion.

File: cmd/kubeadm/app/cmd/phases/controlplanepreparejoin.go
Comment (ereslibre) [reply cmd/kubeadm/app/cmd/phases/controlplanepreparejoin.go]:
```
vagrant@worker:~$ sudo kubeadm join phase control-plane-prepare certs 172.28.128.5:6443
[discovery.bootstrapToken: Invalid value: "": using token-based discovery without caCertHashes can be unsafe. Set unsafeSkipCAVerification to continue, discovery.bootstrapToken.token: Invalid value: "": the bootstrap token is invalid, discovery.tlsBootstrapToken: Invalid value: "": the bootstrap token is invalid]
```

**Review by fabriziopandini — CHANGES_REQUESTED** [review]
@ereslibre The PR cleaned up nicely. Thanks

The last bit to be addressed before merging is the check etcd phase; I don't think this phase should be visible to the users (and in fact this fase was not included in the original list of phases).

So, considering this cannot be moved in prefilght, my proposal is to make ckeck etcd an hidden phase to be executed after control-plane prepare and before kubelet-start. Opinions?

**Comment by ereslibre** [comment]
> So, considering this cannot be moved in prefilght, my proposal is to make ckeck etcd an hidden phase to be executed after control-plane prepare and before kubelet-start. Opinions?

:+1:

**Review by fabriziopandini — APPROVED** [review]
I'm ok with this version. Thanks @ereslibre 
Might be in a follow up we should prefer `checketcd` to `waitetcd` (`checketcd` seems more appropriate to me), but this is a nit not blocking
/approve
Leaving final lgtm to @neolit123 or someone else

**Comment by ereslibre** [comment]
@fabriziopandini Renamed and squashed all commits. Thanks!

**Review by neolit123 — APPROVED** [review]
this looks good, thanks for the update @ereslibre 
/lgtm

**Pull request merged**
